pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/PartialUpsertTableRebalanceIntegrationTest.java

```diff
@@ -0,0 +1,371 @@
+        if (value == null) {
+          return instanceId;
+        } else {
+          assertEquals(instanceId, value);
+          return value;
+        }
+      });
+    }
+
+    assertEquals(uniqueServers.size(), numInstancesExpected);
+  }
+
+  protected void verifyIdealState(int numSegmentsExpected, int numInstancesExpected) {
+    IdealState idealState = HelixHelper.getTableIdealState(_helixManager, REALTIME_TABLE_NAME);
+    verifySegmentAssignment(idealState.getRecord().getMapFields(), numSegmentsExpected, numInstancesExpected);
+  }
+
+  protected void populateTables()
+      throws Exception {
+    // Create and upload segments
+    ClusterIntegrationTestUtils.buildSegmentsFromAvro(_avroFiles, _tableConfig, _schema, 0, _segmentDir, _tarDir);
+    uploadSegments(getTableName(), TableType.REALTIME, _tarDir);
+
+    pushAvroIntoKafka(_avroFiles);
+    // Wait for all documents loaded
+    waitForAllDocsLoaded(600_000L);
+  }
+
+  protected void createSchemaAndTable()
+      throws Exception {
+    // Unpack the Avro files
+    _avroFiles = unpackAvroData(_tempDir);
+
+    // Create and upload schema and table config
+    _schema = createSchema();
+    addSchema(_schema);
+    _tableConfig = createUpsertTableConfig(_avroFiles.get(0), PRIMARY_KEY_COL, null, getNumKafkaPartitions());
+    _tableConfig.getValidationConfig().setDeletedSegmentsRetentionPeriod(null);
+
+    addTableConfig(_tableConfig);
+  }
+
+  @AfterClass
+  public void tearDown() {
+    stopServer();
+    stopBroker();
+    stopController();
+    stopZk();
+  }
+
+  @Override
+  protected String getSchemaFileName() {
+    return "upsert_upload_segment_test.schema";
+  }
+
+  @Override
+  protected String getSchemaName() {
+    return "upsertSchema";
+  }
+
+  @Override
+  protected String getAvroTarFileName() {
+    return "upsert_upload_segment_test.tar.gz";
+  }
+
+  @Override
+  protected String getPartitionColumn() {
+    return PRIMARY_KEY_COL;
+  }
+
+  @Override
+  protected long getCountStarResult() {
+    // Three distinct records are expected with pk values of 100000, 100001, 100002
+    return 3;
+  }
+
+  protected void waitForAllDocsLoaded(long timeoutMs, long expectedCount)
+      throws Exception {
+    TestUtils.waitForCondition(aVoid -> {
+      try {
+        return getCurrentCountStarResultWithoutUpsert() == expectedCount;
+      } catch (Exception e) {
+        return null;
+      }
+    }, 100L, timeoutMs, "Failed to load all documents");
+  }
+
+  private static int getSegmentPartitionId(String segmentName) {
+    switch (segmentName) {
+      case UPLOADED_SEGMENT_1:
+        return 0;
+      case UPLOADED_SEGMENT_2:
+      case UPLOADED_SEGMENT_3:
+        return 1;
+      default:
+        return new LLCSegmentName(segmentName).getPartitionGroupId();
+    }
+  }
+
+  @Override
+  protected void waitForAllDocsLoaded(long timeoutMs)
+      throws Exception {
+    Thread.sleep(5000L); //TODO: Wait for some time for rebalance to kickoff
```

Comment:
We want to avoid sleeping a fixed amount of time. For fast machine, it can waste time; for slow machine, it can cause flakiness. Any specific reason why waiting for condition itself is not enough?

Reply:
Yeah, waiting on condition itself triggers an early termination since the rebalance simply hasn't started and all segments are still available.